### PR TITLE
Reland "Add --serve-observatory flag to run, attach, and test (#118402)" 

### PR DESCRIPTION
**Pull request by bkonyi**
This hopefully fixes the flakiness in `flutter test should respect --serve-observatory` (see https://github.com/flutter/flutter/issues/119722 for details).

### CODE REVIEW

File: packages/flutter_tools/test/integration.shard/test_test.dart [hunk 1]
```diff
@@ -220,6 +219,7 @@ void main() {
       final String content = await response.transform(utf8.decoder).join();
       expect(content.contains('Dart VM Observatory'), true);
     } finally {
+      sub.cancel();
```
Comment (christopherfujino) [inline packages/flutter_tools/test/integration.shard/test_test.dart:222]: what is the code path that we'll reach this point without having ever hit the line where the `sub.cancel()` used to be?

Comment (bkonyi) [reply packages/flutter_tools/test/integration.shard/test_test.dart]: I'll be honest with you, I'm not sure why but this seems to fix the flaky failures. I can run the test with this patch > 50 times under load with no failures, but will see multiple failures in 20 runs with the original patch.

For whatever reason, it seems that the target test process shuts down unexpectedly, but I haven't been able to get any sort of logging to confirm what's going on.

File: packages/flutter_tools/test/integration.shard/test_test.dart [hunk 2]
```diff
@@ -198,6 +198,31 @@ void main() {
...
+
+  testWithoutContext('flutter test should respect --serve-observatory', () async {
+    late final Process process;
+    late final StreamSubscription<String> sub;
+    try {
+      process = await _runFlutterTestConcurrent('trivial', automatedTestsDirectory, flutterTestDirectory,
+        extraArguments: const <String>['--start-paused', '--serve-observatory']);
+      final Completer<Uri> completer = Completer<Uri>();
+      final RegExp vmServiceUriRegExp = RegExp(r'((http)?:\/\/)[^\s]+');
+      sub = process.stdout.transform(utf8.decoder).listen((String e) {
```
Comment (christopherfujino) [inline packages/flutter_tools/test/integration.shard/test_test.dart:210]: @bkonyi

Now I'm not a stream expert, but could it be non-deterministic how long the strings are that get emitted from the `process.stdout` stream? should we put a `LineSplitter` before the `.listen`?

Comment (bkonyi) [reply packages/flutter_tools/test/integration.shard/test_test.dart:210]: I'm not aware of any situations where the `stdout` stream has events that split the middle of a string (we use this same pattern all throughout the SDK test suites).

Comment (christopherfujino) [reply packages/flutter_tools/test/integration.shard/test_test.dart:210]: As far as a process is concerned, it's just sending bytes to a file descriptor, though, right? not strings.

Comment (bkonyi) [reply packages/flutter_tools/test/integration.shard/test_test.dart:210]: Yeah, most likely. This is definitely a common pattern when scraping stdout of a child process.

**Comment by bkonyi** [comment]
Friendly bump. How do we want to move forward with this?

**Comment by christopherfujino** [comment]
> Friendly bump. How do we want to move forward with this?

let me try to poke this to see if I can find the root cause, otherwise I'll approve as is...

**Review by christopherfujino — APPROVED** [review]
You were right, this looks like the right change to make :)

**Comment by bkonyi** [comment]
Let's give this another shot then 🤞

**Pull request merged**